### PR TITLE
DEV: Refactor bed class

### PR DESCRIPTION
**Pull request by thomasyu888**
* use pandas assert frame equals
* Make create_gene_panel function just do one thing
* Remove the need for downloading a file for testing which reduces test time

### CODE REVIEW

**Comment by kdaily** [comment]
Let's do this one in person.

File: genie/bed.py [hunk 1]
```diff
@@ -65,15 +66,14 @@ def create_gtf(dirname):
...
+    """
     exon_gtf_path = os.path.join(dirname, "exon.gtf")
     gene_gtf_path = os.path.join(dirname, "gene.gtf")
 
     if not os.path.exists(exon_gtf_path) or not os.path.exists(gene_gtf_path):
-        download_cmd = [
-            'wget',
-            'http://ftp.ensembl.org/pub/release-75/gtf/homo_sapiens/Homo_sapiens.GRCh37.75.gtf.gz',
-            '-P', dirname]
+        download_cmd = ['wget',
```
Comment (kdaily) [inline genie/bed.py:74]: use synapse to download (file an issue)

File: genie/bed.py [hunk 2]
```diff
@@ -103,22 +103,21 @@ def _add_feature_type_todf(filepath, featuretype):
...
+    """
     # No need to add anything if the dataframe is empty
     if os.stat(filepath).st_size != 0:
-        df = pd.read_csv(filepath, sep="\t", header=None)
-        df.columns = [
-            "Chromosome", "Start_Position", "End_Position",
-            "Hugo_Symbol", "includeInPanel", "clinicalReported",
-            "ID", "SEQ_ASSAY_ID"]
-        df['Feature_Type'] = featuretype
+        beddf = pd.read_csv(filepath, sep="\t", header=None)
```
Comment (kdaily) [inline genie/bed.py:109]: `beddf` is incongruent with function name - consider changing one or the other.

File: genie/bed.py [hunk 3]
```diff
@@ -128,64 +127,56 @@ def add_feature_type(temp_bed_path, exon_gtf_path, gene_gtf_path):
...
-        '|', 'sort', '|', 'uniq', '>', gene_path]
+    command = ['bedtools', 'intersect', '-a',
+               temp_bed_path, '-b', gene_gtf_path, '-wa',
+               '|', 'sort', '|', 'uniq', '>', gene_path]
     subprocess.check_call(" ".join(command), shell=True)
-    command = [
-        'diff', gene_path, genie_exon_path, '|',
-        'grep', "'<'", '|', 'sed', "'s/< //'", '>',
-        genie_intron_path]
+    command = ['diff', gene_path, genie_exon_path, '|',
```
Comment (kdaily) [inline genie/bed.py:162]: what does 'diff' do here? please document.

File: genie/bed.py [hunk 4]
```diff
@@ -128,64 +127,56 @@ def add_feature_type(temp_bed_path, exon_gtf_path, gene_gtf_path):
...
-    # genie_combineddf.sort_values()
-    genie_combined_path = \
-        os.path.join(process_functions.SCRIPT_DIR, "genie_combined.bed")
-    genie_combineddf.to_csv(genie_combined_path, sep="\t", index=False)
-    return(genie_combined_path)
+    return genie_combineddf
 
 
-def _check_to_map(x, gene_positiondf):
+def _check_to_map(row, gene_positiondf):
```
Comment (kdaily) [inline genie/bed.py]: change arg doc from `x` to `row`

File: genie/bed.py [hunk 5]
```diff
@@ -197,131 +188,144 @@ def _check_to_map(x, gene_positiondf):
...
+    return None
+
+
+def _map_gene_within_boundary(row, gene_positiondf, boundary=0.9):
+    """
+    Map gene as long as the strand is a percent within the boundary
+    of the start goes over start boundary, but end is contained in gene
+
+    Args:
+        row: start and end position
```
Comment (kdaily) [inline genie/bed.py]: row has more columns than specified in docstring

File: genie/bed.py [hunk 6]
```diff
@@ -197,131 +188,144 @@ def _check_to_map(x, gene_positiondf):
...
+    """
+    ratio_overlap = overlap / bed_length
+    ratio_overlap = ratio_overlap[ratio_overlap > boundary]
+    ratio_overlap = ratio_overlap[ratio_overlap <= 1]
+    if not ratio_overlap.empty:
+        return ratio_overlap.idxmax()
+    return None
+
+
+def _map_gene_within_boundary(row, gene_positiondf, boundary=0.9):
```
Comment (kdaily) [inline genie/bed.py]: nothing inherent to do with genes - rename?

File: genie/bed.py [hunk 7]
```diff
@@ -197,131 +188,144 @@ def _check_to_map(x, gene_positiondf):
...
+        return_mappeddf: Return mapped dataframe. Default is False.
 
     Return:
         bool or Series: if the gene passed in need to be remapped or
                         the remapped gene
-    '''
+    """
     valid = True
-    to_map = _check_to_map(x, genePositionDf)
+    to_map = _check_to_map(row, gene_positiondf)
```
Comment (kdaily) [inline genie/bed.py]: document what this is doing here

File: genie/bed.py (hunk 4 above)
Comment (kdaily) [reply genie/bed.py]: add more descriptive docs about what 'map' means

File: genie/bed.py [hunk 8]
```diff
@@ -197,131 +188,144 @@ def _check_to_map(x, gene_positiondf):
     Return:
         bool:  Whether mapping needs to happen
     '''
-    match_with_genedb = \
-        gene_positiondf[gene_positiondf['hgnc_symbol'] == x['Hugo_Symbol']]
+    matching_symbol_ind = gene_positiondf['hgnc_symbol'] == row['Hugo_Symbol']
+    match_with_genedb = gene_positiondf[matching_symbol_ind]
```
Comment (kdaily) [inline genie/bed.py:204]: should assert that `match_with_genedb` is only one row, otherwise fail.

File: genie/bed.py [hunk 9]
```diff
@@ -197,131 +188,144 @@ def _check_to_map(x, gene_positiondf):
...
 
     if not match_with_genedb.empty:
         start_position = match_with_genedb['start_position'].values[0]
         end_position = match_with_genedb['end_position'].values[0]
-        if (start_position >= x['Start_Position'] and
-                end_position <= x['End_Position']) or \
-            (start_position <= x['End_Position'] and
-                end_position >= x['Start_Position']):
+        if (start_position >= row['Start_Position'] and
```
Comment (kdaily) [inline genie/bed.py]: document these two cases - hard to determine what has going on

Comment (kdaily) [reply genie/bed.py]: change comparison to work with one unit (either the gene or the submitted row) fixed - that would make the logic more clear

File: genie/bed.py [hunk 10]
```diff
@@ -197,131 +188,144 @@ def _check_to_map(x, gene_positiondf):
...
     if to_map:
-        endRows = _map_gene_within_boundary(x, genePositionDf)
-        if len(endRows) == 0:
-            logger.warning(
-                "{} cannot be remapped. "
-                "These rows will have an empty gene symbol".format(
-                    x['Hugo_Symbol']))
-            x['Hugo_Symbol'] = pd.np.nan
+        end_rows = _map_gene_within_boundary(row, gene_positiondf)
+        if len(end_rows) == 0:
```
Comment (kdaily) [inline genie/bed.py]: test if empty

File: genie/bed.py [hunk 11]
```diff
@@ -197,131 +188,144 @@ def _check_to_map(x, gene_positiondf):
...
+            max_overlap = _get_max_overlap_index(overlap, bed_length,
+                                                 boundary)
+            if max_overlap is not None:
+                end_rows = end_rows.append(chrom_rows.loc[[max_overlap]])
+    return end_rows
 
 
-def validateSymbol(x, genePositionDf, returnMappedDf=False):
-    '''
+def validate_symbol(row, gene_positiondf, return_mappeddf=False):
```
Comment (kdaily) [inline genie/bed.py]: should always return `row` now!

File: genie/bed.py [hunk 12]
```diff
@@ -360,16 +364,15 @@ def _validateFilename(self, filePath):
         assert os.path.basename(filePath[0]).startswith("%s-" % self.center)\
             and os.path.basename(filePath[0]).endswith(".bed")
 
-    def createdBEDandGenePanel(
-            self, bed, seq_assay_id,
-            genePanelPath, parentId,
-            exon_gtf_path,
-            createGenePanel=True):
+    def create_gene_panel(self, temp_bed_path, seq_assay_id,
+                          genePanelPath, parentId,
```
Comment (kdaily) [inline genie/bed.py]: need moar linting!

File: genie/bed.py [hunk 13]
```diff
@@ -379,78 +382,53 @@ def createdBEDandGenePanel(
...
             genie_exon.write(genie_exon_text)
 
         genie_exon_stat = os.stat(genie_exon_path)
         if genie_exon_stat.st_size > 0 and createGenePanel:
-            temp = pd.read_csv(genie_exon_path, sep="\t", header=None)
-            temp.columns = [
-                "Chromosome", "Start_Position", "End_Position",
-                "Hugo_Symbol", "includeInPanel", "clinicalReported",
-                "ID", "SEQ_ASSAY_ID"]
+            exonsdf = pd.read_csv(genie_exon_path, sep="\t", header=None)
```
Comment (kdaily) [inline genie/bed.py]: don't need to write to disk, just read string directly to pandas

File: genie/bed.py [hunk 14]
```diff
@@ -480,19 +458,40 @@ def _process(
...
-            add_feature_type(temp_bed_path, exon_gtf_path, gene_gtf_path)
+        genePosition = self.syn.tableQuery('SELECT * FROM syn11806563')
+        genePositionDf = genePosition.asDataFrame()
+        bed['ID'] = bed['Hugo_Symbol']
+        bed = bed.apply(lambda x: validate_symbol(x, genePositionDf,
+                                                  return_mappeddf=True), axis=1)
+        bed['SEQ_ASSAY_ID'] = seq_assay_id
+        temp_bed_path = os.path.join(process_functions.SCRIPT_DIR, "temp.bed")
+        bed.to_csv(temp_bed_path, sep="\t", index=False, header=None)
+        self.create_gene_panel(temp_bed_path, seq_assay_id,
```
Comment (kdaily) [inline genie/bed.py]: this should return something and not do a `syn.store` inside - you already noted that in 489 you are duplicating work.

**Review by kdaily — CHANGES_REQUESTED** [review]

File: genie/bed.py (hunk 1 above)
Comment (thomasyu888) [reply genie/bed.py:74]: #201

File: genie/bed.py (hunk 13 above)
Comment (thomasyu888) [reply genie/bed.py]: Resolved with #203

File: genie/bed.py (hunk 14 above)
Comment (thomasyu888) [reply genie/bed.py]: Resolve with #203

File: genie/bed.py [hunk 15]
```diff
@@ -128,200 +127,212 @@ def add_feature_type(temp_bed_path, exon_gtf_path, gene_gtf_path):
...
 
 
-def _map_gene_within_boundary(x, genePositionDf, boundary=0.9):
-    '''
-    As long as the strand is a percent within the boundary of the
-    start goes over start boundary, but end is contained in gene
+def _map_position_within_boundary(row, positiondf, boundary=0.9):
+    """
+    Map positions and is contained within the boundary
+    of the start goes over start boundary, but end is contained in gene
```
Comment (thomasyu888) [inline genie/bed.py]: Edit doc

File: genie/bed.py [hunk 16]
```diff
@@ -128,200 +127,212 @@ def add_feature_type(temp_bed_path, exon_gtf_path, gene_gtf_path):
...
-                "These rows will have an empty gene symbol".format(
-                    x['Hugo_Symbol']))
-            x['Hugo_Symbol'] = pd.np.nan
-            valid = False
-        elif len(endRows) > 1:
-            if x['Hugo_Symbol'] not in endRows['hgnc_symbol'].tolist():
+    """
+    region_overlap = _check_region_overlap(row, gene_positiondf)
+    if not region_overlap:
+        overlap_genes = _map_position_within_boundary(row, gene_positiondf)
```
Comment (thomasyu888) [inline genie/bed.py]: named `overlap_positions`

**Review by kdaily — APPROVED** [review]
“as soon as we renounce fiction and illusion, we lose reality itself; the moment we subtract fictions from reality, reality itself loses its discursive-logical consistency.” - Slavoj Žižek

This PR get's the @kdaily stamp of approval.

**Pull request merged**